pkg/assembler/nodes.go

```diff
@@ -16,9 +16,161 @@
+
+func (an AttestationNode) Properties() map[string]interface{} {
+	properties := make(map[string]interface{})
+	properties["file"] = an.File
+	properties["checksum"] = an.Checksum
+	return properties
+}
+
+func (an AttestationNode) Attributes() []string {
+	return []string{"file", "checksum"}
+}
+
+func (an AttestationNode) IdentifiableAttributes() [][]string {
+	// An attestation can be uniquely identified by filename?
+	return [][]string{{"file"}}
+}
+
+// BuilderNode is a node that represents an attestation
+type BuilderNode struct {
+	BuilderType string
+	BuilderId   string
+}
+
+func (bn BuilderNode) Type() string {
+	return "Builder"
+}
+
+func (bn BuilderNode) Properties() map[string]interface{} {
+	properties := make(map[string]interface{})
+	properties["type"] = bn.BuilderType
+	properties["id"] = bn.BuilderId
+	return properties
+}
+
+func (bn BuilderNode) Attributes() []string {
+	return []string{"type", "id"}
+}
+
+func (bn BuilderNode) IdentifiableAttributes() [][]string {
+	// A builder needs both type and id to be identified
+	return [][]string{{"type", "id"}}
+}
+
+// AttestationEdge is an edge that represents the fact that an
+// `AttestationNode` is an attestation for an `ArtifactNode`.
+type AttestationEdge struct {
+	AttestationNode AttestationNode
+	ArtifactNode    ArtifactNode
+}
+
+func (e AttestationEdge) Type() string {
+	return "Attestation"
+}
+
+func (e AttestationEdge) Nodes() (v, u GuacNode) {
+	return e.AttestationNode, e.ArtifactNode
+}
+
+func (e AttestationEdge) Properties() map[string]interface{} {
+	return map[string]interface{}{}
+}
+
+func (e AttestationEdge) Attributes() []string {
+	return []string{}
+}
+
+func (e AttestationEdge) IdentifiableAttributes() [][]string {
+	return [][]string{}
+}
+
+// BuiltByEdge is an edge that represents the fact that an
+// `ArtifactNode` has been built by a `BuilderNode`
+type BuiltByEdge struct {
+	ArtifactNode ArtifactNode
+	BuilderNode  BuilderNode
+}
+
+func (e BuiltByEdge) Type() string {
+	return "BuiltBy"
+}
+
+func (e BuiltByEdge) Nodes() (v, u GuacNode) {
+	return e.ArtifactNode, e.BuilderNode
+}
+
+func (e BuiltByEdge) Properties() map[string]interface{} {
+	return map[string]interface{}{}
+}
+
+func (e BuiltByEdge) Attributes() []string {
+	return []string{}
+}
+
+func (e BuiltByEdge) IdentifiableAttributes() [][]string {
+	return [][]string{}
+}
+
+// DependsOnEdge is an edge that represents the fact that an
+// `ArtifactNode` depends on another `ArtifactNode`
+type DependsOnEdge struct {
+	ArtifactNode ArtifactNode
+	Dependency   ArtifactNode
+}
+
+func (e DependsOnEdge) Type() string {
+	return "DependsOn"
+}
+
+func (e DependsOnEdge) Nodes() (v, u GuacNode) {
+	return e.ArtifactNode, e.Dependency
+}
+
+func (e DependsOnEdge) Properties() map[string]interface{} {
+	return map[string]interface{}{}
+}
+
+func (e DependsOnEdge) Attributes() []string {
+	return []string{}
+}
+
+func (e DependsOnEdge) IdentifiableAttributes() [][]string {
+	return [][]string{}
```

Comment:
Wouldn't the combination of `ArtifactNode` and `Dependency` be uniquely identifying? I'm not convinced but it's a pity to missing a primary key (even a synthetic one if necessary).

Reply:
It would be, but for edges this function allows for other ways to uniquely identify them.

For example, if we support a query of "for all attestations from this time window return all artifacts" if `AttestationForEdge` has the date as an attribute (instead of being stored in the attestation node) we could use that date to select the edge and then from there select the artifacts.

It is very likely that by the time we have an end-to-end example we realize that we don't need other ways to identify an edge except by its endpoints, at which point I'd remove this.